It is not possible to compile RustPython (0.5.0 and current main) with feature `ssl-vendor`

## Summary

```text
# cargo clean && cargo build --release --no-default-features --features threading,stdlib,stdio,importlib,host_env,freeze-stdlib,sqlite,ssl-vendor
     Removed 3814 files, 1.1GiB total
  Downloaded openssl v0.10.76
  Downloaded openssl-sys v0.9.112
  Downloaded 2 crates (355.0KiB) in 0.54s
   Compiling proc-macro2 v1.0.106
   Compiling unicode-ident v1.0.22
   Compiling quote v1.0.45
...
   Compiling dirs-next v2.0.0
   Compiling lexopt v0.3.2
   Compiling env_logger v0.11.10
   Compiling liblzma v0.4.6
error[E0425]: cannot find type `LazyLock` in this scope
   --> crates/stdlib/src/openssl.rs:435:24
    |
435 |     static CERT_PATHS: LazyLock<(PathBuf, PathBuf)> = LazyLock::new(|| {
    |                        ^^^^^^^^ not found in this scope
    |
help: consider importing one of these structs
    |
 53 +     use crate::openssl::LazyLock;
    |
 53 +     use std::sync::LazyLock;
    |

error[E0433]: cannot find type `LazyLock` in this scope
   --> crates/stdlib/src/openssl.rs:435:55
    |
435 |     static CERT_PATHS: LazyLock<(PathBuf, PathBuf)> = LazyLock::new(|| {
    |                                                       ^^^^^^^^ use of undeclared type `LazyLock`
    |
help: consider importing one of these structs
    |
 53 +     use crate::openssl::LazyLock;
    |
 53 +     use std::sync::LazyLock;
    |

error[E0425]: cannot find type `LazyLock` in this scope
   --> crates/stdlib/src/openssl.rs:473:28
    |
473 |     static CERT_ENV_NAMES: LazyLock<(String, String)> = LazyLock::new(|| {
    |                            ^^^^^^^^ not found in this scope
    |
help: consider importing one of these structs
    |
 53 +     use crate::openssl::LazyLock;
    |
 53 +     use std::sync::LazyLock;
    |

error[E0433]: cannot find type `LazyLock` in this scope
   --> crates/stdlib/src/openssl.rs:473:57
    |
473 |     static CERT_ENV_NAMES: LazyLock<(String, String)> = LazyLock::new(|| {
    |                                                         ^^^^^^^^ use of undeclared type `LazyLock`
    |
help: consider importing one of these structs
    |
 53 +     use crate::openssl::LazyLock;
    |
 53 +     use std::sync::LazyLock;
    |

error[E0599]: no method named `as_str` found for struct `PyRef<PyStr>` in the current scope
    --> crates/stdlib/src/openssl.rs:1049:38
     |
1049 |             let ciphers = cipherlist.as_str();
     |                                      ^^^^^^
     |
help: there is a method `as_pystr` with a similar name, but with different arguments
    --> crates/vm/src/builtins/str.rs:256:5
     |
 256 |     fn as_pystr(self, ctx: &Context) -> &'a Py<PyStr>;
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0599]: no method named `as_str` found for struct `PyRef<PyStr>` in the current scope
    --> crates/stdlib/src/openssl.rs:1105:26
     |
1105 |                     if s.as_str().contains('\0') {
     |                          ^^^^^^
     |
help: there is a method `as_pystr` with a similar name, but with different arguments
    --> crates/vm/src/builtins/str.rs:256:5
     |
 256 |     fn as_pystr(self, ctx: &Context) -> &'a Py<PyStr>;
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0599]: no method named `as_str` found for struct `PyRef<PyStr>` in the current scope
    --> crates/stdlib/src/openssl.rs:1462:64
     |
1462 |                     *self.psk_identity_hint.lock() = Some(hint.as_str().to_owned());
     |                                                                ^^^^^^
     |
help: there is a method `as_pystr` with a similar name, but with different arguments
    --> crates/vm/src/builtins/str.rs:256:5
     |
 256 |     fn as_pystr(self, ctx: &Context) -> &'a Py<PyStr>;
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0599]: no method named `as_str` found for struct `PyRef<PyStr>` in the current scope
    --> crates/stdlib/src/openssl.rs:1490:31
     |
1490 |                         if !s.as_str().is_ascii() {
     |                               ^^^^^^
     |
help: there is a method `as_pystr` with a similar name, but with different arguments
    --> crates/vm/src/builtins/str.rs:256:5
     |
 256 |     fn as_pystr(self, ctx: &Context) -> &'a Py<PyStr>;
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0599]: no method named `as_str` found for reference `&Py<PyStr>` in the current scope
    --> crates/stdlib/src/openssl.rs:1980:22
     |
1980 |                 Ok(s.as_str().as_bytes().to_vec())
     |                      ^^^^^^
     |
help: there is a method `as_pystr` with a similar name, but with different arguments
    --> crates/vm/src/builtins/str.rs:256:5
     |
 256 |     fn as_pystr(self, ctx: &Context) -> &'a Py<PyStr>;
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0599]: no method named `as_str` found for reference `&PyRef<PyStr>` in the current scope
    --> crates/stdlib/src/openssl.rs:2036:45
     |
2036 |                 let hostname_str = hostname.as_str();
     |                                             ^^^^^^
     |
help: there is a method `as_pystr` with a similar name, but with different arguments
    --> crates/vm/src/builtins/str.rs:256:5
     |
 256 |     fn as_pystr(self, ctx: &Context) -> &'a Py<PyStr>;
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0599]: no method named `as_str` found for reference `&PyRef<PyStr>` in the current scope
    --> crates/stdlib/src/openssl.rs:2755:75
     |
2755 |             let cb_type_str = cb_type.as_ref().map_or("tls-unique", |s| s.as_str());
     |                                                                           ^^^^^^
     |
help: there is a method `as_pystr` with a similar name, but with different arguments
    --> crates/vm/src/builtins/str.rs:256:5
     |
 256 |     fn as_pystr(self, ctx: &Context) -> &'a Py<PyStr>;
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Some errors have detailed explanations: E0425, E0433, E0599.
For more information about an error, try `rustc --explain E0425`.
error: could not compile `rustpython-stdlib` (lib) due to 11 previous errors
```

No such problem when compiling with `ssl-rustls` instead of `ssl-vendor`.

The last commit where `ssl-vendor` compiles is c06cf56c60db1ddf34e4c2037edd6153fb21e7e1. Compilation is broken since at least ccd3d4f964e7c9139b99c7d1154e5eb6e764d8cc.

## Expected

Successful compilation.

## Actual

Compilation errors.

## Python Documentation

N/A

--- a/.github/workflows/ci.yaml
+++ b/.github/workflows/ci.yaml
@@ -140,6 +140,10 @@ jobs:
         run: cargo build --no-default-features --features ssl-openssl
         if: runner.os == 'Linux'
 
+      - name: Test vendored OpenSSL build
+        run: cargo build --no-default-features --features ssl-vendor
+        if: runner.os == 'Linux'
+
       # - name: Install tk-dev for tkinter build
       #   run: sudo apt-get update && sudo apt-get install -y tk-dev
       #   if: runner.os == 'Linux'

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -217,6 +217,7 @@ exitcode = "1.1.2"
 flame = "0.2.2"
 flamer = "0.5"
 flate2 = { version = "1.1.9", default-features = false }
+# Bump only when the openssl crate bumps it
 foreign-types-shared = "0.1"
 gethostname = "1.0.2"
 getrandom = { version = "0.3", features = ["std"] }

--- a/crates/stdlib/src/openssl.rs
+++ b/crates/stdlib/src/openssl.rs
@@ -65,7 +65,7 @@ mod _ssl {
             LazyLock, PyMappedRwLockReadGuard, PyMutex, PyRwLock, PyRwLockReadGuard,
             PyRwLockWriteGuard,
         },
-        socket::{self, PySocket},
+        socket::{self, PySocket, SockWaitKind, sock_wait},
         vm::{
             AsObject, Py, PyObject, PyObjectRef, PyPayload, PyRef, PyResult, VirtualMachine,
             builtins::{
@@ -2297,34 +2297,31 @@ mod _ssl {
             needs: SslNeeds,
             deadline: &SocketDeadline,
             vm: &VirtualMachine,
-        ) -> SelectRet {
-            let sock = match self.0.sock_opt() {
-                Some(s) => s,
-                None => return SelectRet::Closed,
+        ) -> PyResult<SelectRet> {
+            let Some(sock) = self.0.sock_opt() else {
+                return Ok(SelectRet::Closed);
             };
-            // For blocking sockets without timeout, call sock_wait with None timeout
+            // For blocking sockets without timeout, call sock_select with None timeout
             // to actually block waiting for data instead of busy-looping
             let timeout = match &deadline {
                 Ok(deadline) => match deadline.checked_duration_since(Instant::now()) {
                     Some(d) => Some(d),
-                    None => return SelectRet::TimedOut,
+                    None => return Ok(SelectRet::TimedOut),
                 },
                 Err(true) => None, // Blocking: no timeout, wait indefinitely
-                Err(false) => return SelectRet::Nonblocking,
+                Err(false) => return Ok(SelectRet::Nonblocking),
             };
-            let res = socket::sock_wait(
-                &sock,
-                match needs {
-                    SslNeeds::Read => socket::SockWaitKind::Read,
-                    SslNeeds::Write => socket::SockWaitKind::Write,
-                },
-                timeout,
-                vm,
-            );
-            match res {
-                Ok(true) => SelectRet::TimedOut,
-                _ => SelectRet::Ok,
-            }
+            let wait_kind = match needs {
+                SslNeeds::Read => SockWaitKind::Read,
+                SslNeeds::Write => SockWaitKind::Write,
+            };
+            sock_wait(&*sock, wait_kind, timeout, vm).map(|timed_out| {
+                if timed_out {
+                    SelectRet::TimedOut
+                } else {
+                    SelectRet::Ok
+                }
+            })
         }
 
         fn socket_needs(
@@ -2332,14 +2329,15 @@ mod _ssl {
             err: &ssl::Error,
             deadline: &SocketDeadline,
             vm: &VirtualMachine,
-        ) -> (Option<SslNeeds>, SelectRet) {
+        ) -> PyResult<(Option<SslNeeds>, SelectRet)> {
             let needs = match err.code() {
                 ssl::ErrorCode::WANT_READ => Some(SslNeeds::Read),
                 ssl::ErrorCode::WANT_WRITE => Some(SslNeeds::Write),
                 _ => None,
             };
-            let state = needs.map_or(SelectRet::Ok, |needs| self.select(needs, deadline, vm));
-            (needs, state)
+            let state =
+                needs.map_or(Ok(SelectRet::Ok), |needs| self.select(needs, deadline, vm))?;
+            Ok((needs, state))
         }
     }
 
@@ -2857,7 +2855,7 @@ mod _ssl {
                         break;
                     }
                     // Wait briefly for peer's close_notify before retrying
-                    match socket_stream.select(SslNeeds::Read, &deadline, vm) {
+                    match socket_stream.select(SslNeeds::Read, &deadline, vm)? {
                         SelectRet::TimedOut => {
                             return Err(socket::timeout_error_msg(
                                 vm,
@@ -2895,7 +2893,7 @@ mod _ssl {
                 };
 
                 // Wait on the socket
-                match socket_stream.select(needs, &deadline, vm) {
+                match socket_stream.select(needs, &deadline, vm)? {
                     SelectRet::TimedOut => {
                         let msg = if err == sys::SSL_ERROR_WANT_READ {
                             "The read operation timed out"
@@ -2991,7 +2989,7 @@ mod _ssl {
                 let (needs, state) = stream
                     .get_ref()
                     .expect("handshake called in bio mode; should only be called in socket mode")
-                    .socket_needs(&err, &timeout, vm);
+                    .socket_needs(&err, &timeout, vm)?;
                 match state {
                     SelectRet::TimedOut => {
                         // Clean up SNI ex_data before returning error
@@ -3045,7 +3043,7 @@ mod _ssl {
                 .get_ref()
                 .expect("write called in bio mode; should only be called in socket mode");
             let timeout = socket_ref.timeout_deadline();
-            let state = socket_ref.select(SslNeeds::Write, &timeout, vm);
+            let state = socket_ref.select(SslNeeds::Write, &timeout, vm)?;
             match state {
                 SelectRet::TimedOut => {
                     return Err(socket::timeout_error_msg(
@@ -3065,7 +3063,7 @@ mod _ssl {
                 let (needs, state) = stream
                     .get_ref()
                     .expect("write called in bio mode; should only be called in socket mode")
-                    .socket_needs(&err, &timeout, vm);
+                    .socket_needs(&err, &timeout, vm)?;
                 match state {
                     SelectRet::TimedOut => {
                         return Err(socket::timeout_error_msg(
@@ -3236,7 +3234,7 @@ mod _ssl {
                     let (needs, state) = stream
                         .get_ref()
                         .expect("read called in bio mode; should only be called in socket mode")
-                        .socket_needs(&err, &timeout, vm);
+                        .socket_needs(&err, &timeout, vm)?;
                     match state {
                         SelectRet::TimedOut => {
                             return Err(socket::timeout_error_msg(
@@ -4149,7 +4147,7 @@ mod bio {
     use openssl_sys as sys;
     use std::marker::PhantomData;
 
-    pub struct MemBioSlice<'a>(*mut sys::BIO, PhantomData<&'a [u8]>);
+    pub(super) struct MemBioSlice<'a>(*mut sys::BIO, PhantomData<&'a [u8]>);
 
     impl Drop for MemBioSlice<'_> {
         fn drop(&mut self) {
@@ -4160,7 +4158,7 @@ mod bio {
     }
 
     impl<'a> MemBioSlice<'a> {
-        pub fn new(buf: &'a [u8]) -> Result<MemBioSlice<'a>, ErrorStack> {
+        pub(super) fn new(buf: &'a [u8]) -> Result<MemBioSlice<'a>, ErrorStack> {
             openssl::init();
 
             assert!(buf.len() <= c_int::MAX as usize);
@@ -4172,7 +4170,7 @@ mod bio {
             Ok(MemBioSlice(bio, PhantomData))
         }
 
-        pub fn as_ptr(&self) -> *mut sys::BIO {
+        pub(super) fn as_ptr(&self) -> *mut sys::BIO {
             self.0
         }
     }

--- a/crates/stdlib/src/ssl/error.rs
+++ b/crates/stdlib/src/ssl/error.rs
@@ -125,6 +125,10 @@ pub(crate) mod ssl_error {
         )
     }
 
+    #[cfg_attr(
+        all(feature = "ssl-openssl", not(feature = "ssl-rustls")),
+        expect(dead_code)
+    )]
     pub(crate) fn create_ssl_zero_return_error(vm: &VirtualMachine) -> PyRef<PyOSError> {
         vm.new_os_subtype_error(
             PySSLZeroReturnError::class(&vm.ctx).to_owned(),
@@ -133,6 +137,10 @@ pub(crate) mod ssl_error {
         )
     }
 
+    #[cfg_attr(
+        all(feature = "ssl-openssl", not(feature = "ssl-rustls")),
+        expect(dead_code)
+    )]
     pub(crate) fn create_ssl_syscall_error(
         vm: &VirtualMachine,
         msg: impl Into<String>,

